File: Andrea/stepminer.m

%x1 - xn, n is time points (vector of increasing values)
%x is values (vector)
%X is resulting vector

function [u, threshold] = stepminer(x, alpha)

n = length(x);
%initialize
%setting is twostep function true or false
[SSR, SSE, t1, t2] = SSCalculate(n,x);
P1 = pval(SSE(1), SSR(1), n, 3);
P2 = pval(SSE(2), SSR(2), n, 4);
P12 = F12(SSE(1),SSE(2), 3, 4, n);

%determine which to use better

if P1<alpha % && (P12 > alpha)
    %type = "oneStep";
    u = x > t1;
    threshold = t1
elseif P2<alpha
    %type = "twoStep";
    u = x > t2;
    threshold = t2
else
    %type = "other";
    u = zeros(1,n);
    threshold = NaN
    
end


end


function [SSR, SSE, threshold1, threshold2] = SSCalculate(n, x)  %n here is the length not the vector

    %Calculate SSTOT

    xmean = mean(x);

    SSTOT = sum((x(1:n)-xmean).^2);

    %find means from left and right as it goes
    %and find smallest SSE for optimal step
    %considering all possibilities

    %rightMeanMin = rightMean
    %leftMeanMin = leftMean
    %breakIndex = -1

    SSEmin1 = SSTOT;
    SSEmin2 = SSTOT;

    %SSR = SSTOT - SSE
    %SSE = SSTOT - SSR
    %find min SSE

    for i = 1:n-1
        %add up mean from left side
        leftMean = mean(x(1:i));
        %add up mean from right side
        % this is so it doesnt divide by 0
        rightMean = mean(x(i+1:n));
        if i < n-1
            %all values from breakpoint i until possible breakpoints
            for j = i+1:n-1

                rightMean2 = mean(x(i+1:j));
                leftMean2 = (mean(x(1:i))*(i) + mean(x(j+1:n)) * (n-j))/(n-j+i);

                %first amount before breakpoint * leftMean
                %amount since breakpoint1 to breakpoint2
                %amount from breakpoint2 to end
                SSE2 = sum((x(1:i)-leftMean2).^2) + sum((x(i+1:j)-rightMean2).^2) + sum((x(j+1:n)-leftMean2).^2);

                if SSEmin2 > SSE2
                    SSEmin2 = SSE2;

                    %binarized vector
                    %values

                    threshold2 = mean([leftMean2 rightMean2]);

                    %breakpoints2(1) = leftMean2
                    %breakpoints2(2) = rightMean2
                    %breakpoints2(3) = i
                    %breakpoints2(4) = j

                end
                
            end


        %calculate SSE
        %calculate SSR first since xmean is fixed
        %and fixed values are constant until break point
    

        SSE1 = sum((x(1:i)-leftMean).^2) + sum((x(i+1:n)-rightMean).^2);
   
        %try every break point until min SSE
        if SSEmin1 > SSE1
            SSEmin1 = SSE1;

            %store means and breakpoints

            threshold1 = mean([leftMean rightMean]);

            %breakpoints1(1) = leftMean
            %breakpoints1(2) = rightMean
            %breakpoints1(3) = i
    
        end
    end

    %last return values
    SSE1 = SSEmin1;
    SSR1 = SSTOT - SSE1;

    SSE2 = SSEmin2;
    SSR2 = SSTOT - SSE2;

    SSR = [SSR1, SSR2];
    SSE = [SSE1, SSE2];

end


%F

function [P, m] = pval (SSE, SSR, n, m)



    MSR = SSR/(m-1);

    %MSE

    MSE = SSE/(n-m);

    %F

    F = MSR/MSE;

    %f distribution probability
    % i think this is correct
    %p value

    P = fpdf(F,m-1,n-m);


end

%F12

function F = F12(SSE1, SSE2, m1, m2, n)
    F = abs(((SSE1-SSE2)/(m2-m1))/(SSE2/(n-m2)));
    F = chi2cdf(F,1);
end

File: Andrea/stepminer.py
#This stepminer works as the paper describes, it verifies if the F value given
#for each optimal step function is greater than the critical F for the degrees
#of freedom for the F distribution based on the number of elements in the 
#vector. If F is below the critical value for both then the final vector will
#just result in 0 as neither onestep nor twostep are enough to reject the
#null hypothesis.


import numpy as np
import scipy.stats

#FUNCTIONS


def stepminer(x, alpha):
    n = len(x)
    
    
    SSE, SSR, t1, t2 = SScalculate(n,x)
    
    
    t = None
    
    #print(SSE)
    #print(SSR)
    
    m = 3
    
    #onestep
    MSR = SSR[0]/(m-1)
    MSE = SSE[0]/(n-m)
    F = MSR/MSE
    
    if F > scipy.stats.f.ppf(q=1-alpha, dfn=m-1, dfd=n-m):
        t = t1
        #print("step 1")
    else:      
         m = 4
         
         #twostep
         MSR = SSR[1]/(m-1)
         MSE = SSE[1]/(n-m)
         F = MSR/MSE
         if F > scipy.stats.f.ppf(q=1-alpha, dfn=m-1, dfd=n-m):
             t = t2
             #("step 2")
         #else:
            #print("no reg")
             

    #initialize at 0
             
    u = np.zeros(n)
    
    #change above threshold to 1's
    if t:
        for i in range(n):
            if x[i] >= t:
                u[i] = 1
          
    #return vector and threshold
    return u, t
    
def SScalculate(n, x):
    
    xmean = np.mean(x)
    SSTOT = getSSTOT(x, n, xmean)
    
    SSEmin1 = SSTOT
    SSEmin2 = SSTOT
    
    
    for i in range(n - 1):
        leftMean = np.mean(x[0:i+1])
    
        rightMean = np.mean(x[i+1:n])
        
        if i < n-2:
            for j in range(i+1, n-1):
                
                #verify later
                rightMean2 = np.mean(x[i+1:j+1])
                leftMean2 = (leftMean*(i+1) + np.mean(x[j+1:n])*(n-j-1))/(n-j+i)
                
                
                SSE2 = 0
                
                for k in range(n):
                    if k < i+1 or k > j:
                        SSE2 = SSE2 + (x[k]-leftMean2)**2
                    else:
                        SSE2 = SSE2 + (x[k] - rightMean2)**2
                        
                if SSEmin2 > SSE2:
                    SSEmin2 = SSE2
                    #print("2:",SSEmin2)
                    
                    t2 = (leftMean2 + rightMean2)/2
                    
        SSE1 = 0
        for j in range(n):
            if j < i+1:
                SSE1 = SSE1 + (x[j] - leftMean)**2
            else:
                SSE1 = SSE1 + (x[j] - rightMean)**2
                
        if SSEmin1 > SSE1:
            SSEmin1 = SSE1
            #print("1:",SSEmin1)
            
            t1 = (leftMean + rightMean)/2
            
            
    
    SSE = np.array([SSEmin1, SSEmin2])
    SSR = np.array([SSTOT - SSEmin1, SSTOT - SSEmin2])
    
    return SSE, SSR, t1, t2





    
def getSSTOT(x, n, xmean):
    m = 0
    for i in range(n):
        m = m + (x[i] - xmean)**2
    return m


#TESTING

HIVdat = np.loadtxt("HIVIn(Matlab).csv",delimiter=",")
blank = np.empty([len(HIVdat), len(HIVdat[0])])

for i in range(len(HIVdat)):
    arr, t = stepminer(HIVdat[i], 0.05)
    blank[i] = arr

np.savetxt("stepminerpy.txt", blank, delimiter=",", fmt='%d')